openfl/federated/task/runner_xgb.py

```diff
@@ -0,0 +1,382 @@
+            and input_tensor_dict["local_tree"].size != 0
+        ) or (
+            not isinstance(input_tensor_dict["local_tree"], np.ndarray)
+            and input_tensor_dict["local_tree"] is not None
+        ):
+            self.set_tensor_dict(input_tensor_dict)
+
+    def validate_task(self, col_name, round_num, input_tensor_dict, **kwargs):
+        """Validate Task.
+
+        Run validation of the model on the local data.
+
+        Args:
+            col_name (str): Name of the collaborator.
+            round_num (int): What round is it.
+            input_tensor_dict (dict): Required input tensors (for model).
+            **kwargs: Additional parameters.
+
+        Returns:
+            global_output_dict (dict):  Tensors to send back to the aggregator.
+            local_output_dict (dict):   Tensors to maintain in the local
+                TensorDB.
+        """
+        loader = self.data_loader.get_valid_dmatrix()
+
+        # during agg validation, self.bst will still be None. during local validation, it will have a value - no need to rebuild
+        if self.bst is None:
+            self.rebuild_model(input_tensor_dict)
+
+        # if self.bst is still None after rebuilding, then there was no initial global model, so set metric to 0
+        if self.bst is None:
+            # for first round agg validation, there is no model so set metric to 0
+            # TODO: this is not robust, especially if using a loss metric
+            metric = Metric(name="accuracy", value=np.array(0))
+        else:
+            metric = self.validate_(loader)
+
+        origin = col_name
+        suffix = "validate"
+        if kwargs["apply"] == "local":
+            suffix += "_local"
+        else:
+            suffix += "_agg"
+        tags = ("metric",)
+        tags = change_tags(tags, add_field=suffix)
+        # TODO figure out a better way to pass in metric for this pytorch
+        #  validate function
+        output_tensor_dict = {TensorKey(metric.name, origin, round_num, True, tags): metric.value}
+
+        # Empty list represents metrics that should only be stored locally
+        return output_tensor_dict, {}
+
+    def train_task(
+        self,
+        col_name,
+        round_num,
+        input_tensor_dict,
+        **kwargs,
+    ):
+        """Train batches task.
+
+        Train the model on the requested number of batches.
+
+        Args:
+            col_name (str): Name of the collaborator.
+            round_num (int): What round is it.
+            input_tensor_dict (dict): Required input tensors (for model).
+            **kwargs: Additional parameters.
+
+        Returns:
+            global_output_dict (dict):  Tensors to send back to the aggregator.
+            local_output_dict (dict):   Tensors to maintain in the local
+                TensorDB.
+        """
+        self.rebuild_model(input_tensor_dict)
+        loader = self.data_loader.get_train_dmatrix()
```

Comment:
Same as above, suggesting to rename `loader` to `data` for more seamless reading.